proving/worker.go

```diff
@@ -0,0 +1,193 @@
+package proving
+
+import (
+	"bytes"
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"fmt"
+	"io"
+	"sort"
+	"sync"
+	"unsafe"
+
+	"github.com/spacemeshos/post/oracle"
+	"github.com/spacemeshos/post/shared"
+)
+
+type batch struct {
+	Data    []byte
+	Index   uint64
+	Release func()
+}
+
+type solution struct {
+	Nonce uint
+	Index uint64
+}
+
+// ioWorker is a worker that reads labels from disk and writes them to a batch channel to be processed by the
+// labelWorkers.
+//
+// TODO(mafa): use this as base to replace GranSpecificReader / GranSpecificWriter and the persistence package.
+func ioWorker(ctx context.Context, batchChan chan<- *batch, b uint32, source io.ReadCloser) error {
+	defer close(batchChan)
+	defer source.Close()
+	index := uint64(0)
+
+	batchDataPool := sync.Pool{
+		New: func() any {
+			buf := make([]byte, BlocksPerWorker*b)
+			return &buf
+		},
+	}
+
+	for {
+		data := *(batchDataPool.Get().(*[]byte))
+		n, err := source.Read(data)
+		switch {
+		case err == io.EOF || err == nil:
+			// make sure we don't send partial blocks to the label workers.
+			// TODO(mafa): this silently drops partial blocks and should be handled better.
+			n -= n % int(b)
+			batch := &batch{
+				Data:  data[:n],
+				Index: index,
+				Release: func() {
+					batchDataPool.Put(&data)
+				},
+			}
+			select {
+			case batchChan <- batch:
+				if err == io.EOF {
+					return nil
+				}
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		default:
+			data = data[:n]
+			batchDataPool.Put(&data)
+			return err
+		}
+		index += uint64(n) / aes.BlockSize
+	}
+}
+
+// labelWorker is a worker that receives batches from ioWorker and looks for indices to be included in the proof.
+func labelWorker(ctx context.Context, batchChan <-chan *batch, solutionChan chan<- *solution, ch shared.Challenge, numOuts uint8, numNonces uint32, b uint32, d uint, difficulty uint64) error {
+	// use two slices with different types that point to the same memory location.
+	// this is done to speed up the conversation from bytes to uint64.
+	out := make([]byte, numOuts*aes.BlockSize+8)
+	u64s := make([][]uint64, 8)
+	for i := range u64s {
+		size := (len(out) - i) / 8
+		u64s[i] = unsafe.Slice((*uint64)(unsafe.Pointer(&out[i])), size)
+	}
```

Comment:
It casts `[]byte` -> `[][]uint64`. For this to be safe, the `out []byte` should probably be 8B-aligned. Are `[]byte` guaranteed to be 8B-aligned in Go?

Casting in the opposite direction seems safer to me.

Reply:
I'm not sure I can follow `[]byte` isn't guaranteed to be 8B aligned, why would that matter in this approach?

Let's say the memory looks like this:

```
offset:  00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
out:              b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 bA bB bC
u64s[0]:          u0 u0 u0 u0 u0 u0 u0 u0 u1 u1 u1 u1 u1
u64s[1]:             u0 u0 u0 u0 u0 u0 u0 u0 u1 u1 u1 u1
u64s[2]:                u0 u0 u0 u0 u0 u0 u0 u0 u1 u1 u1
...
```

The start of `u64[i]` always points at `out[i]`. I don't know what you mean by casting `uint64` to `byte`? I still need to cast `byte` to `uint64` for all other slices besides the one that is used to cast `[]byte` to?

Reply:
I'm not sure about Go. Although in C/C++ it matters. A pointer to uint64 should be aligned to 8B in order for the CPU to load it efficiently into registers. Otherwise, it could do it inefficiently or even crash (https://stackoverflow.com/questions/1496848/does-unaligned-memory-access-always-cause-bus-errors/1496881#1496881).


By
> casts `[]byte` -> `[][]uint64`
I meant that you reinterpret the memory under `[]byte` as `[][]uint64` (so that the u64 slices point to the underlying byte array).

I checked on my PC and accessing bytes from `out` as `uint64` in an unaligned way _works_ so it probably is fine on x86.